### PR TITLE
ENT-4978 Add BillingProducer to worker component

### PR DESCRIPTION
**Pull request by mstead**
https://issues.redhat.com/browse/ENT-4978

* Add secondary listener to tally summary topic.
* Implemented a bare bones BillingProducer class that listens
  to the TallySummary topic and logs that a message was recieved
  and then forwards to new billable usage topic:
     platform.rhsm-subscriptions.billable-usage

## Testing
1. Ensure you have some Events in the DB and take not of the message counts for each topic.
2. Start the app as follows:
```bash
LOGGING_LEVEL_ORG_CANDLEPIN_SUBSCRIPTIONS_TALLY_BILLING=DEBUG SUBSCRIPTION_USE_STUB=true USER_USE_STUB=true DEV_MODE=true  ./gradlew :bootRun
```
3. Run the hourly tally. BE SURE YOU TALLY AROUND THE RANGE THAT HAS EVENTS. The example below should cover any Events in may and force a fresh re-tally.
```bash
curl 'http://localhost:8080/actuator/hawtio/jolokia/' \
-H 'Content-Type: application/json' \
-d '{"type":"exec","mbean":"org.candlepin.subscriptions.jmx:name=tallyJmxBean,type=TallyJmxBean","operation":"tallyAccountByHourly(java.lang.String,java.lang.String,java.lang.String)","arguments":["account123", "2022-05-01T00:00Z", "2022-05-30T00:00Z"]}'
```
4. Verify that you see logs from the RhMarketplaceMapper which means we are still getting TallySummary messages to the RHM consumer (likely subscription not found type messages).
```bash
2022-05-13 16:55:30,704 [thread=rh-marketplace-worker-0-C-1] [ERROR] [org.candlepin.subscriptions.rhmarketplace.RhMarketplacePayloadMapper] - SUBSCRIPTIONS3001: Could not find marketplace subscription id
```

5. Verify that you see DEBUG logs stating that we are getting messages to the secondary TallySummary consumer
```bash
2022-05-13 16:55:30,595 [thread=billing-producer-tally-summary-consumers-0-C-1] [DEBUG] [org.candlepin.subscriptions.tally.billing.TallySummaryMessageConsumer] - Tally Summary recieved. Producing billable usage.}
2022-05-13 16:55:30,595 [thread=billing-producer-tally-summary-consumers-0-C-1] [DEBUG] [org.candlepin.subscriptions.tally.billing.BillingProducer] - Forwarding summary account123 to topic platform.rhsm-subscriptions.billable-usage
```
8. Verify the  platform.rhsm-subscriptions.billable-usage topic exists and has messages waiting to be consumed.

### CODE REVIEW

File: src/main/java/org/candlepin/subscriptions/tally/billing/BillingProducer.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.kafka.core.KafkaTemplate;
+import org.springframework.stereotype.Service;
+
+/**
+ * Component that produces billing usage messages based on TallySnapshots.
+ *
+ * <p>NOTE: We are currently just forwarding TallySummary messages, but will transition to sending
+ * BillingUsage.
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/tally/billing/BillingProducer.java:35]:
```suggestion
 * BillingUsage.
```
```suggestion
 * BillableUsage.
```

File: src/main/java/org/candlepin/subscriptions/tally/billing/BillingProducer.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ */
+@Service
+@Slf4j
+public class BillingProducer {
+
+  private KafkaTemplate<String, TallySummary> billingUsageKafkaTemplate;
+  private String billingUsageTopic;
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/tally/billing/BillingProducer.java:42]:
```suggestion
  private String billingUsageTopic;
```
```suggestion
  private String billableUsageTopic;
```

File: src/main/java/org/candlepin/subscriptions/tally/billing/BillingProducer.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+  @Autowired
+  public BillingProducer(
+      @Qualifier("billingUsageTopicProperties") TaskQueueProperties billingUsageTopicProperties,
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/tally/billing/BillingProducer.java:46]:
```suggestion
      @Qualifier("billingUsageTopicProperties") TaskQueueProperties billingUsageTopicProperties,
```
```suggestion
      @Qualifier("billableUsageTopicProperties") TaskQueueProperties billableUsageTopicProperties,
```

File: src/main/java/org/candlepin/subscriptions/tally/billing/BillingProducer.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      @Qualifier("billingUsageKafkaTemplate")
+          KafkaTemplate<String, TallySummary> billingUsageKafkaTemplate) {
+    this.billingUsageKafkaTemplate = billingUsageKafkaTemplate;
+    this.billingUsageTopic = billingUsageTopicProperties.getTopic();
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/tally/billing/BillingProducer.java:50]:
```suggestion
      @Qualifier("billableUsageTopicProperties") TaskQueueProperties billableUsageTopicProperties,
      @Qualifier("billableUsageKafkaTemplate")
          KafkaTemplate<String, TallySummary> billableUsageKafkaTemplate) {
    this.billableUsageKafkaTemplate = billableUsageKafkaTemplate;
    this.billableUsageTopic = billableUsageTopicProperties.getTopic();
```

Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/tally/billing/BillingProducer.java:50]:
```suggestion
      @Qualifier("billableUsageTopicProperties") TaskQueueProperties billableUsageTopicProperties,
      @Qualifier("billableUsageKafkaTemplate")
          KafkaTemplate<String, TallySummary> billableUsageKafkaTemplate) {
    this.billableUsageKafkaTemplate = billableUsageKafkaTemplate;
    this.billableUsageTopic = billableUsageTopicProperties.getTopic();
```

File: src/main/java/org/candlepin/subscriptions/tally/billing/BillingProducer.java (hunk 2 above)
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/tally/billing/BillingProducer.java:42]:
```suggestion
  private KafkaTemplate<String, TallySummary> billableUsageKafkaTemplate;
  private String billableUsageTopic;
```

Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/tally/billing/BillingProducer.java:42]:
```suggestion
  private KafkaTemplate<String, TallySummary> billableUsageKafkaTemplate;
  private String billableUsageTopic;
```

File: src/main/java/org/candlepin/subscriptions/tally/billing/BillingProducer.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  public void produce(TallySummary tallySummary) {
+    log.debug(
+        "Forwarding summary {} to topic {}",
+        tallySummary.getAccountNumber(),
+        this.billingUsageTopic);
+    billingUsageKafkaTemplate.send(this.billingUsageTopic, tallySummary);
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/tally/billing/BillingProducer.java:58]:
```suggestion
        this.billableUsageTopic);
    billableUsageKafkaTemplate.send(this.billableUsageTopic, tallySummary);
```

Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/tally/billing/BillingProducer.java:58]:
```suggestion
        this.billableUsageTopic);
    billableUsageKafkaTemplate.send(this.billableUsageTopic, tallySummary);
```

File: src/main/java/org/candlepin/subscriptions/tally/billing/BillingProducerConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  }
+
+  @Bean
+  @ConfigurationProperties(prefix = "rhsm-subscriptions.billing-producer.messaging.billable-usage")
+  public TaskQueueProperties billingUsageTopicProperties() {
+    return new TaskQueueProperties();
+  }
+
+  @Bean
+  public ProducerFactory<String, TallySummary> billingUsageProducerFactory(
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/tally/billing/BillingProducerConfiguration.java:112]:
```suggestion
  public ProducerFactory<String, TallySummary> billableUsageProducerFactory(
```

File: src/main/java/org/candlepin/subscriptions/tally/billing/BillingProducerConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+     */
+    factory.setValueSerializer(new JsonSerializer<>(objectMapper));
+    return factory;
+  }
+
+  @Bean
+  public KafkaTemplate<String, TallySummary> billingUsageKafkaTemplate(
+      @Qualifier("billingUsageProducerFactory")
+          ProducerFactory<String, TallySummary> billingUsageProducerFactory) {
+    return new KafkaTemplate<>(billingUsageProducerFactory);
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/tally/billing/BillingProducerConfiguration.java:128]:
```suggestion
  public KafkaTemplate<String, TallySummary> billingUsageKafkaTemplate(
      @Qualifier("billingUsageProducerFactory")
          ProducerFactory<String, TallySummary> billingUsageProducerFactory) {
    return new KafkaTemplate<>(billingUsageProducerFactory);
```
```suggestion
  public KafkaTemplate<String, TallySummary> billableUsageKafkaTemplate(
      @Qualifier("billableUsageProducerFactory")
          ProducerFactory<String, TallySummary> billableUsageProducerFactory) {
    return new KafkaTemplate<>(billableUsageProducerFactory);
```

File: src/main/java/org/candlepin/subscriptions/tally/billing/BillingProducerConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+          .setIdleEventInterval(kafkaProperties.getListener().getIdleEventInterval().toMillis());
+    }
+    // hack to track the Kafka consumers, so SeekableKafkaConsumer can commit when needed
+    factory.getContainerProperties().setConsumerRebalanceListener(registry);
+    return factory;
+  }
+
+  @Bean
+  @ConfigurationProperties(prefix = "rhsm-subscriptions.billing-producer.messaging.billable-usage")
+  public TaskQueueProperties billingUsageTopicProperties() {
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/tally/billing/BillingProducerConfiguration.java:107]:
```suggestion
  public TaskQueueProperties billableUsageTopicProperties() {
```

**Review by kahowell — CHANGES_REQUESTED** [review]
Looks good overall, just looking for s/billingUsage/billableUsage/ in a few places. I think I marked them all, but please sanity check with `git grep -i billingUsage` afterwards.

File: src/main/resources/application-worker.yaml [hunk 9]
```diff
@@ -35,3 +35,14 @@ rhsm-subscriptions:
...
     max-attempts: ${TALLY_SUMMARY_PRODUCER_MAX_ATTEMPTS:5}
+  billing-producer:
+    back-off-initial-interval: ${BILLING_PRODUCER_BACK_OFF_INITIAL_INTERVAL:1s}
+    back-off-max-interval: ${BILLING_PRODUCER_BACK_OFF_MAX_INTERVAL:1m}
+    back-off-multiplier: ${BILLING_PRODUCER_BACK_OFF_MULTIPLIER:2}
+    max-attempts: ${BILLING_PRODUCER_MAX_ATTEMPTS:1}
+    messaging:
+      tally-summary:
+        topic: platform.rhsm-subscriptions.tally
+        kafka-group-id: billing-producer-tally-summary-consumers
```
Comment (kahowell) [inline src/main/resources/application-worker.yaml:46]:
```suggestion
        kafka-group-id: swatch-producer-billing
```

**Review by kahowell — CHANGES_REQUESTED** [review]

File: src/main/resources/application-worker.yaml [hunk 10]
```diff
@@ -35,3 +35,15 @@ rhsm-subscriptions:
...
+  billing-producer:
+    back-off-initial-interval: ${BILLING_PRODUCER_BACK_OFF_INITIAL_INTERVAL:1s}
+    back-off-max-interval: ${BILLING_PRODUCER_BACK_OFF_MAX_INTERVAL:1m}
+    back-off-multiplier: ${BILLING_PRODUCER_BACK_OFF_MULTIPLIER:2}
+    max-attempts: ${BILLING_PRODUCER_MAX_ATTEMPTS:1}
+    incoming:
+        topic: platform.rhsm-subscriptions.tally
+        kafka-group-id: swatch-producer-billing
+        seek-override-end: ${KAFKA_SEEK_OVERRIDE_END:false}
+        seek-override-timestamp: ${KAFKA_SEEK_OVERRIDE_TIMESTAMP:}
```
Comment (kahowell) [inline src/main/resources/application-worker.yaml:47]: nitpick: too many spaces; please use 2 less:
```suggestion
      topic: platform.rhsm-subscriptions.tally
      kafka-group-id: swatch-producer-billing
      seek-override-end: ${KAFKA_SEEK_OVERRIDE_END:false}
      seek-override-timestamp: ${KAFKA_SEEK_OVERRIDE_TIMESTAMP:}
```

**Comment by mstead** [comment]
Now sending BillableUsage messages to billable-usage topic and and fixed the formatting.

Should be the last thing needed here.

**Review by kahowell — APPROVED** [review]

**Pull request merged**